### PR TITLE
BuildStatus moved into sdks model + realtime updates

### PR DESCRIPTION
**Pull request by PatrickShaw**
## Proposed changes
State managers now update entities in real time. In order to get this feature done the following code changes had to be made:
- Replaced all map field names in state managers with `entities`.
- Added a method makes the service in a state manager listen to real time updates
- Moved BuildStatus into the `sdk` model. This probably could have been done in a different PR but doing both real time updates and the build status stuff is just.
Most file changes are pretty superficial. Just some variable renaming.

## Types of changes

What types of changes does your code introduce to OpenAPI Platform?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I have signed and submitted the appropriate [CLA](../docs/contrib)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/frontend-dist/src/state/withRealtimeUpdates.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+export function withRealtimeUpdates<T extends Timestamped, S extends CrudState<T>>(
+  state: S,
+): S {
+  const addIfNewerEntities = action((entity: HasId<T>) => {
+    if (incomingIsNewer(state.entities.get(entity.id), entity)) {
+      state.entities.set(entity.id, entity);
+    }
+  });
+  ['created', 'patched', 'updated'].forEach(event => {
+    state.service.on(event, addIfNewerEntities);
```
Comment (calebj0seph) [inline packages/frontend-dist/src/state/withRealtimeUpdates.tsx]: Are you sure the patched event gives you the full object and not just part of it?

File: packages/frontend-dist/src/state/incomingIsNewer.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * @returns True if the incoming entity is newer
+ */
+export function incomingIsNewer<T extends Timestamped>(
+  original: T | undefined,
+  incoming: T,
+): boolean {
+  if (!original) {
+    return true;
+  } else if (original.updatedAt && incoming.updatedAt) {
+    if (original.updatedAt < incoming.updatedAt) {
```
Comment (calebj0seph) [inline packages/frontend-dist/src/state/incomingIsNewer.tsx]: Just make this `return original.updatedAt < incoming.updatedAt;`

File: packages/frontend-dist/src/state/incomingIsNewer.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      return true;
+    } else {
+      return false;
+    }
+  } else if (original.updatedAt) {
+    return false;
+  } else if (incoming.updatedAt) {
+    return true;
+  } else if (original.createdAt < incoming.createdAt) {
```
Comment (calebj0seph) [inline packages/frontend-dist/src/state/incomingIsNewer.tsx]: Just make this
```
} else {
  return original.createdAt < incoming.createdAt;
}
```

File: packages/frontend-dist/src/view/basic/SdkConfigItem.tsx [hunk 4]
```diff
@@ -43,24 +49,63 @@ export interface SdkConfigItemProps extends React.DOMAttributes<HTMLDivElement>
...
+  constructor(props) {
+    super(props);
+    state.retrieveLatestSdk(this.props.sdkConfig);
+  }
+
+  @computed
+  public get latestSdk() {
+    const i = state.entities.values();
+    let currentSdk: (Sdk & Partial<PathHolder>) | null = null;
+    for (const sdk of i) {
```
Comment (calebj0seph) [inline packages/frontend-dist/src/view/basic/SdkConfigItem.tsx]: Wouldn't `i.find(...);` be way more concise than a huge for loop?

File: packages/frontend-dist/src/state/withRealtimeUpdates.tsx (hunk 1 above)
Comment (PatrickShaw) [reply packages/frontend-dist/src/state/withRealtimeUpdates.tsx]: Yep, I've checked. But also, the whole PR would straight up just not work if that were the case.

File: packages/frontend-dist/src/state/incomingIsNewer.tsx (hunk 2 above)
Comment (PatrickShaw) [reply packages/frontend-dist/src/state/incomingIsNewer.tsx]: True, originally I thought updatedAt would be null but it never is.

File: packages/frontend-dist/src/view/basic/SdkConfigItem.tsx (hunk 4 above)
Comment (PatrickShaw) [reply packages/frontend-dist/src/view/basic/SdkConfigItem.tsx]: I didn't know that was a method in iterators. Ty :)

Comment (PatrickShaw) [reply packages/frontend-dist/src/view/basic/SdkConfigItem.tsx]: From what I can see, it's not a thing actually. I suppose I could do `Array.from(...)`. Also, I believe you meant `reduce` as opposed to `find`

**Pull request merged**